test_rmw_implementation/test/test_subscription.cpp

```diff
@@ -554,47 +563,63 @@ TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), ignore_local_publicat
   original_message.bool_value = true;
   original_message.char_value = 'k';
   original_message.float32_value = 3.14159f;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    test_msgs__msg__BasicTypes__fini(&original_message);
+  });
+
   rmw_publisher_allocation_t * null_allocation_p{nullptr};
   rmw_subscription_allocation_t * null_allocation_s{nullptr};
 
-  // ignore_local_publications = true
+  ret = rmw_publish(pub, &original_message, null_allocation_p);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+  rmw_subscriptions_t subscriptions;
+  void * subscriptions_storage[2];
+  subscriptions_storage[0] = sub_ignorelocal->data;
+  subscriptions_storage[1] = sub->data;
+  subscriptions.subscribers = subscriptions_storage;
+  subscriptions.subscriber_count = 2;
+
+  rmw_wait_set_t * wait_set = rmw_create_wait_set(&context, 2);
+  ASSERT_NE(nullptr, wait_set) << rmw_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
   {
-    ret = rmw_publish(pub, &original_message, null_allocation_p);
-    EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+    EXPECT_EQ(
+      RMW_RET_OK, rmw_destroy_wait_set(wait_set)) << rmw_get_error_string().str;
+  });
+  rmw_time_t timeout = {1, 0};  // 1000ms
+  ret = rmw_wait(&subscriptions, nullptr, nullptr, nullptr, nullptr, wait_set, &timeout);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
```

Comment:
@jamoralp consider checking the state of `subscriptions` to make sure both received the message (as described in  [`rmw_wait`](https://github.com/ros2/rmw/blob/a50f3bb1e24ac5338677990d74ab096a9d1776d7/rmw/include/rmw/rmw.h#L2425-L2429) documentation).

Reply:
This will fail for CycloneDDS RMW, since the `ignore_local_publication` flag for their middleware basically blocks the reception of DDS message at a middleware level, not at RMW level; thus, the sub_ignorelocal subscription condition will never be triggered for their RMW. I will check the second subscriber to not bu NULL though.